src/app.ts

```diff
@@ -90,16 +90,26 @@ export class App extends Construct {
     // the necessary operations. We do however want to preserve the distributed validation.
     validate(this);
 
-    const simpleManifestNamer = (chart: Chart) => `${Names.toDnsLabel(chart)}.k8s.yaml`;
-    const manifestNamer = hasDependantCharts ? (chart: Chart) => `${index.toString().padStart(4, '0')}-${simpleManifestNamer(chart)}` : simpleManifestNamer;
+    const charts = new DependencyGraph(Node.of(this)).topology()
+      .filter(x => x instanceof Chart);
 
-    const charts: IConstruct[] = new DependencyGraph(Node.of(this)).topology().filter(x => x instanceof Chart);
-
-    let index = 0;
+    const found = new Set<IConstruct>();
+    const namer: ChartNamer = hasDependantCharts ? new IndexedChartNamer() : new SimpleChartNamer();
     for (const node of charts) {
       const chart: Chart = Chart.of(node);
-      Yaml.save(path.join(this.outdir, manifestNamer(chart)), chartToKube(chart));
-      index++;
+
+      // charts may nest each other and have overlapping scopes, so objects must be deduped
+      // otherwise two or more charts may include the same object
+      const objects = new DependencyGraph(Node.of(chart)).topology()
+        .filter(x => (x instanceof ApiObject) && !found.has(x));
+
+      const chartName = namer.name(chart);
+      const manifestContents = objects.map(x => (x as ApiObject).toJson());
+      Yaml.save(path.join(this.outdir, chartName), manifestContents);
+
+      for (const obj of objects) {
+        found.add(obj);
+      }
```

Comment:
Initially I was planning to go a step further and impose that if a chart is rendering no options, then it shouldn't bother creating a file/manifest for it (i.e. not just prune objects, but prune files as well) -- but there are currently unit tests that expect e.g. two empty charts to generate two empty files, so I guess the possibility of generating empty manifests is expected behavior.

Reply:
Generating empty is the expected behavior yes, `new Chart` is equivalent to manually create a manifest file.